In [79]:
import sys

print(f"Python version: {sys.version}")
print(f"Platform: {sys.platform}")
print(f"Executable path: {sys.executable}")
print(f"Byte order: {sys.byteorder}")
print(f"Copyright: {sys.copyright}")


Python version: 3.14.0 (main, Oct  7 2025, 16:05:28) [GCC 13.3.0]
Platform: linux
Executable path: /usr/bin/python
Byte order: little
Copyright: Copyright (c) 2001 Python Software Foundation.
All Rights Reserved.

Copyright (c) 2000 BeOpen.com.
All Rights Reserved.

Copyright (c) 1995-2001 Corporation for National Research Initiatives.
All Rights Reserved.

Copyright (c) 1991-1995 Stichting Mathematisch Centrum, Amsterdam.
All Rights Reserved.


In [80]:
import platform

print(f"System: {platform.system()}")
print(f"Node Name: {platform.node()}")
print(f"Release: {platform.release()}")
print(f"Version: {platform.version()}")
print(f"Machine: {platform.machine()}")
print(f"Processor: {platform.processor()}")
print(f"Platform: {platform.platform()}")

System: Linux
Node Name: 1fc6f483796d
Release: 6.8.0-50-generic
Version: #51-Ubuntu SMP PREEMPT_DYNAMIC Sat Nov  9 18:03:35 UTC 2024
Machine: aarch64
Processor: aarch64
Platform: Linux-6.8.0-50-generic-aarch64-with-glibc2.39


In [81]:
print("Hello world")

Hello world


Comments can be placed at the end of a line, and Python will ignore the rest of the line:
A comment does not have to be text that explains the code, it can also be used to prevent Python from executing code:

In [82]:
# This is the single line comment
print("Hello world") # This is a comment

Hello world


Multiline Comments </br>
Python does not really have a syntax for multiline comments.

To add a multiline comment you could insert a # for each line:

In [83]:
# This is a comment
# written in
# more than just one line
print("Hello, World!")

Hello, World!


Or, not quite as intended, you can use a multiline string.

Since Python will ignore string literals that are not assigned to a variable, you can add a multiline string (triple quotes) in your code, and place your comment inside it:

In [84]:
"""
This is a multiline comment.
Python ignores this string if it’s not assigned to anything.
You can write multiple lines here.
"""
print("")

Semicolons are optional in Python. </br>
You can write multiple statements on one line by separating them with ; but this is rarely used because it makes it hard to read:

In [85]:
print("Hello"); print("How are you?"); print("Bye bye!")

Hello
How are you?
Bye bye!


However, if you put two statements on the same line without a separator (newline or ;), Python will give an error:

In [86]:
print("Python is fun!") print("Really!")

SyntaxError: invalid syntax (2531607073.py, line 1)

This is the print example

In [87]:
print("2 + 2 is : "+ str((2+2)))

num1 = num2 = 2
result = num1 + num2
print(f"{num1} + {num2} is : {result}")

print("{} + {} is : {}".format(num1, num2, result))

print(num1, "+", num2, "is :",result)

2 + 2 is : 4
2 + 2 is : 4
2 + 2 is : 4
2 + 2 is : 4


Text in Python must be inside quotes. You can use either " double quotes or ' single quotes: </br>
If you forget to put the text inside quotes, Python will give an error:

In [88]:
print("This will work!")
print('This will also work!')

This will work!
This will also work!


By default, the print() function ends with a new line. </br>
If you want to print multiple words on the same line, you can use the end parameter: </br>
Note that we add a space after end="  " for better readability.

In [89]:
print("Hello World!", end="  ")
print("I will print on the same line.")

Hello World!  I will print on the same line.


You can combine text and numbers in one output by separating them with a comma:

In [90]:
print("I am", 35, "years old.")

I am 35 years old.


In [91]:
x = "Python "
y = "is "
z = "awesome"
print(x + y + z)

Python is awesome


In [92]:
x = 5
y = 10
print(x + y)

15


In [93]:
x = 5
y = "John"
print(x + y)

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [ ]:
x = 5
y = "John"
print(x, y)

Python has no command for declaring a variable.

A variable is created the moment you first assign a value to it.

In [94]:
x = 5
y = "John"

print(x)
print(y)

5
John


Variables do not need to be declared with any particular type, and can even change type after they have been set.

In [95]:
x = 4       # x is of type int
x = "Sally" # x is now of type str
print(x)

Sally


If you want to specify the data type of a variable, this can be done with casting. </br>
You can get the data type of a variable with the type() function.

In [96]:
x = str(3)    # x will be '3'
print(type(x))

y = int(3)    # y will be 3
print(type(y))

z = float(3)  # z will be 3.0
print(type(z))

<class 'str'>
<class 'int'>
<class 'float'>


String variables can be declared either by using single or double quotes:

In [97]:
x = "John"
# is the same as
y = 'John'
print((x == y))

True


Variable names are case-sensitive.

In [98]:
a = 4
A = "Sally"
# A will not overwrite a

A variable can have a short name (like x and y) or a more descriptive name (age, carname, total_volume).

Rules for Python variables:
- A variable name must start with a letter or the underscore character
- A variable name cannot start with a number
- A variable name can only contain alpha-numeric characters and underscores (A-z, 0-9, and _ )
- Variable names are case-sensitive (age, Age and AGE are three different variables)
- A variable name cannot be any of the Python keywords.

Many Values to Multiple Variables

Python allows you to assign values to multiple variables in one line: </br>
Note: Make sure the number of variables matches the number of values, or else you will get an error.

In [99]:
x, y, z = "Orange", "Banana", "Cherry"
print(x)
print(y)
print(z)

Orange
Banana
Cherry


One Value to Multiple Variables

And you can assign the same value to multiple variables in one line:

In [100]:
x = y = z = "Orange"
print(x)
print(y)
print(z)

Orange
Orange
Orange


Unpack a Collection </br>
If you have a collection of values in a list, tuple etc. Python allows you to extract the values into variables. This is called unpacking.

In [101]:
fruits = ["apple", "banana", "cherry"]
x, y, z = fruits
print(x)
print(y)
print(z)

apple
banana
cherry


Global Variables </br>
Variables that are created outside of a function (as in all of the examples in the previous pages) are known as global variables.

Global variables can be used by everyone, both inside of functions and outside.

Create a variable outside of a function, and use it inside the function

In [102]:
x = "awesome"

def myfunc():
  print("Python is " + x)

myfunc()

Python is awesome


If you create a variable with the same name inside a function, this variable will be local, and can only be used inside the function. </br>
The global variable with the same name will remain as it was, global and with the original value.

In [103]:
x = "awesome"

def myfunc():
  x = "fantastic"
  print("(In function)", "Python is " + x)

myfunc()

print("(In global)", "Python is " + x)

(In function) Python is fantastic
(In global) Python is awesome


The global Keyword

Normally, when you create a variable inside a function, that variable is local, and can only be used inside that function.</br>
To create a global variable inside a function, you can use the global keyword.

In [104]:
def myfunc():
  global x
  x = "fantastic"

myfunc()

print("Python is " + x)

Python is fantastic


To change the value of a global variable inside a function, refer to the variable by using the global keyword:

In [105]:
x = "awesome"

def myfunc():
  global x
  x = "fantastic"

print("before calling the function", "Python is " + x)

myfunc()

print("After calling the function", "Python is " + x)

before calling the function Python is awesome
After calling the function Python is fantastic


## Python Data Types

### Python None

None is a special constant in Python that represents the absence of a value.

Its data type is NoneType, and None is the only instance of a NoneType object.

Variables can be assigned None to indicate "no value" or "not set".   

In [179]:
x = None
print(x)
print(type(x))

None
<class 'NoneType'>


To compare a value to None, use the identity operator is or is not

In [180]:
result = None
if result is None:
  print("No result yet")
else:
  print("Result is ready")

No result yet


Similar example, but using is not instead:

In [181]:
result = None
if result is not None:
  print("Result is ready")
else:
  print("No result yet")

No result yet


None evaluates to False in a boolean context.

In [182]:
print(bool(None))

False


Functions that do not explicitly return a value return None by default.

In [185]:
def myfunc():
  x = 5

x = myfunc()
print(x)

None


### Built-in Data Types

In programming, data type is an important concept.

Variables can store data of different types, and different types can do different things.

Python has the following data types built-in by default, in these categories:

- Text Type:	str
- Numeric Types:	int, float, complex
- Sequence Types:	list, tuple, range
- Mapping Type:	dict
- Set Types:	set, frozenset
- Boolean Type:	bool
- Binary Types:	bytes, bytearray, memoryview
- None Type:	NoneType

In Python, the data type is set when you assign a value to a variable:

In [106]:
x = "Hello World"	# str
print(type(x), x)

x = 20	# int	
print(type(x), x)

x = 20.5	# float	
print(type(x), x)

x = 1j	# complex	
print(type(x), x)

x = ["apple", "banana", "cherry"]	# list	
print(type(x), x)

x = ("apple", "banana", "cherry")	# tuple	
print(type(x), x)

x = range(6)	# range	
print(type(x), x)

x = {"name" : "John", "age" : 36}	# dict	
print(type(x), x)

x = {"apple", "banana", "cherry"}	# set	
print(type(x), x)

x = frozenset({"apple", "banana", "cherry"})	# frozenset	
print(type(x), x)

x = True	# bool	
print(type(x), x)

x = b"Hello"	# bytes	
print(type(x), x)

x = bytearray(5)	# bytearray
print(type(x), x)

x = memoryview(bytes(5))	# memoryview	
print(type(x), x)

x = None	# NoneType
print(type(x), x)

<class 'str'> Hello World
<class 'int'> 20
<class 'float'> 20.5
<class 'complex'> 1j
<class 'list'> ['apple', 'banana', 'cherry']
<class 'tuple'> ('apple', 'banana', 'cherry')
<class 'range'> range(0, 6)
<class 'dict'> {'name': 'John', 'age': 36}
<class 'set'> {'banana', 'apple', 'cherry'}
<class 'frozenset'> frozenset({'banana', 'apple', 'cherry'})
<class 'bool'> True
<class 'bytes'> b'Hello'
<class 'bytearray'> bytearray(b'\x00\x00\x00\x00\x00')
<class 'memoryview'> <memory at 0xf48e7ff341c0>
<class 'NoneType'> None


If you want to specify the data type, you can use the following constructor functions:

In [107]:
x = str("Hello World")	# str
print(type(x), x)

x = int(20)	# int	
print(type(x), x)

x = float(20.5)	# float	
print(type(x), x)

x = complex(1j)	# complex	
print(type(x), x)

x = list(("apple", "banana", "cherry"))	# list	
print(type(x), x)

x = tuple(("apple", "banana", "cherry"))	# tuple	
print(type(x), x)

x = range(6)	# range	
print(type(x), x)

x = dict(name="John", age=36)	# dict	
print(type(x), x)

x = set(("apple", "banana", "cherry"))	# set
print(type(x), x)

x = frozenset(("apple", "banana", "cherry"))	# frozenset	
print(type(x), x)

x = bool(5)	# bool	
print(type(x), x)

x = bytes(5)	# bytes	
print(type(x), x)

x = bytearray(5)	# bytearray	
print(type(x), x)

x = memoryview(bytes(5)) # memoryview
print(type(x), x)

<class 'str'> Hello World
<class 'int'> 20
<class 'float'> 20.5
<class 'complex'> 1j
<class 'list'> ['apple', 'banana', 'cherry']
<class 'tuple'> ('apple', 'banana', 'cherry')
<class 'range'> range(0, 6)
<class 'dict'> {'name': 'John', 'age': 36}
<class 'set'> {'banana', 'cherry', 'apple'}
<class 'frozenset'> frozenset({'banana', 'cherry', 'apple'})
<class 'bool'> True
<class 'bytes'> b'\x00\x00\x00\x00\x00'
<class 'bytearray'> bytearray(b'\x00\x00\x00\x00\x00')
<class 'memoryview'> <memory at 0xf48e7ff341c0>


Python Numbers

In [108]:
x = 1    # int
y = 2.8  # float
z = 1j   # complex

print(type(x))
print(type(y))
print(type(z))

<class 'int'>
<class 'float'>
<class 'complex'>


Int, or integer, is a whole number, positive or negative, without decimals, of unlimited length.

In [109]:
x = 1
y = 35656222554887711
z = -3255522

print(type(x))
print(type(y))
print(type(z))

<class 'int'>
<class 'int'>
<class 'int'>


Float, or "floating point number" is a number, positive or negative, containing one or more decimals.

In [110]:
x = 1.10
y = 1.0
z = -35.59

print(type(x))
print(type(y))
print(type(z))

<class 'float'>
<class 'float'>
<class 'float'>


Float can also be scientific numbers with an "e" to indicate the power of 10.

In [111]:
x = 35e3
y = 12E4
z = -87.7e100

print(type(x), x)
print(type(y), y)
print(type(z), z)

<class 'float'> 35000.0
<class 'float'> 120000.0
<class 'float'> -8.77e+101


There may be times when you want to specify a type on to a variable. This can be done with casting. Python is an object-orientated language, and as such it uses classes to define data types, including its primitive types.

Casting in python is therefore done using constructor functions:
- int() - constructs an integer number from an integer literal, a float literal (by removing all decimals), or a string literal (providing the string represents a whole number)
- float() - constructs a float number from an integer literal, a float literal or a string literal (providing the string represents a float or an integer)
- str() - constructs a string from a wide variety of data types, including strings, integer literals and float literals

In [112]:
x = int(1)   # x will be 1
y = int(2.8) # y will be 2
z = int("3") # z will be 3

print(x, y, z)

1 2 3


In [113]:
x = float(1)     # x will be 1.0
y = float(2.8)   # y will be 2.8
z = float("3")   # z will be 3.0
w = float("4.2") # w will be 4.2

print(x, y, z, w)

1.0 2.8 3.0 4.2


In [114]:
x = str("s1") # x will be 's1'
y = str(2)    # y will be '2'
z = str(3.0)  # z will be '3.0'

print(type(x), type(y), type(z))

<class 'str'> <class 'str'> <class 'str'>


## Boolean

Booleans represent one of two values: True or False.

In [115]:
print(10 > 9)
print(10 == 9)
print(10 < 9)

True
False
False


In [116]:
a = 200
b = 33

if b > a:
  print("b is greater than a")
else:
  print("b is not greater than a")

b is not greater than a


The bool() function allows you to evaluate any value, and give you True or False in return,

In [117]:
x = "Hello"
y = 15

print(bool(x))
print(bool(y))

True
True


## Python Operators

### Arithmetic Operators

In [118]:
x = 15
y = 4

print(x + y)
print(x - y)
print(x * y)
print(x / y)
print(x % y)  # Modulus
print(x ** y) # Exponentiation
print(x // y) # Floor division

19
11
60
3.75
3
50625
3


### Assignment Operators

In [119]:
numbers = [1, 2, 3, 4, 5]
count = len(numbers)
if count > 3:
    print(f"List has {count} elements")

if (count := len(numbers)) > 3:
    print(f"List has {count} elements")

List has 5 elements
List has 5 elements


### Comparison Operators

In [120]:
x = 5
y = 3

print(x == y)
print(x != y)
print(x > y)
print(x < y)
print(x >= y)
print(x <= y)

False
True
True
False
True
False


#### Chaining Comparison Operators

In [121]:
x = 5

print(1 < x < 10)

print(1 < x and x < 10)

True
True


### Python Logical Operators

In [122]:
x = 5

print(x > 0 and x < 10)
print(x < 5 or x > 10)
print(not(x > 3 and x < 10))

True
False
False


### Python Identity Operators

The is operator returns True if both variables point to the same object:

In [123]:
x = ["apple", "banana"]
y = ["apple", "banana"]
z = x

print(x is z) # True as pointing to the same object
print(x is y) # False as pointing to different objects
print(x == y) # True as Both have the same content

True
False
True


The is not operator returns True if both variables do not point to the same object:

In [124]:
x = ["apple", "banana"]
y = ["apple", "banana"]

print(x is not y) # True as pointing to different objects

True


Difference Between `is` and `==` </br>
- **is** checks if both variables point to the same object in memory
- **==** checks if the values of both variables are equal

In [125]:
x = [1, 2, 3]
y = [1, 2, 3]

print(x == y) # True as Both have the same content
print(x is y) # False as pointing to different objects

True
False


### Python Membership Operators

Check if "banana" is present in a list:

In [126]:
fruits = ["apple", "banana", "cherry"]

print("banana" in fruits)

True


Check if "pineapple" is NOT present in a list:

In [127]:
fruits = ["apple", "banana", "cherry"]

print("pineapple" not in fruits)

True


The membership operators also work with strings:

In [128]:
text = "Hello World"

print("H" in text)
print("hello" in text)
print("z" not in text)

True
False
True


### Python Bitwise Operators

In [129]:
print(format(6, f'05b'))
print(format(3, f'05b'))
print(6 & 3)

00110
00011
2


In [130]:
print(format(6, f'05b'))
print(format(3, f'05b'))
print(6 | 3)

00110
00011
7


In [131]:
print(format(6, f'05b'))
print(format(3, f'05b'))
print(6 ^ 3)

00110
00011
5


### Python Operator Precedence

#### Operator Precedence
Operator precedence describes the order in which operations are performed.

Parentheses has the highest precedence, meaning that expressions inside parentheses must be evaluated first:

In [132]:
print((6 + 3) - (6 + 3))

0


Multiplication * has higher precedence than addition +, and therefore multiplications are evaluated before additions:

In [133]:
print(100 + 5 * 3)

115


#### Precedence Order

#### Left-to-Right Evaluation

If two operators have the same precedence, the expression is evaluated from left to right.

Addition + and subtraction - has the same precedence, and therefore we evaluate the expression from left to right:

In [134]:
print(5 + 4 - 7 + 3)

5


## Python If Statement

In [147]:
a = 33
b = 200
if b > a:
  print("b is greater than a")

b is greater than a


In this example we use two variables, a and b, which are used as part of the if statement to test whether b is greater than a. As a is 33, and b is 200, we know that 200 is greater than 33, and so we print to screen that "b is greater than a".


### Python Elif Statement

The elif keyword is Python's way of saying "if the previous conditions were not true, then try this condition".

The elif keyword allows you to check multiple expressions for True and execute a block of code as soon as one of the conditions evaluates to True.

Use elif when you have multiple mutually exclusive conditions to check. This is more efficient than using multiple separate if statements because Python stops checking once it finds a true condition.

You can have as many elif statements as you need. Python will check each condition in order and execute the first one that is true.

In [159]:
score = 75

if score >= 90:
  print("Grade: A")
elif score >= 80:
  print("Grade: B")
elif score >= 70:
  print("Grade: C")
elif score >= 60:
  print("Grade: D")

Grade: C


When you use elif, Python evaluates the conditions from top to bottom. As soon as it finds a condition that is true, it executes that block and skips all remaining conditions.

Important: Only the first true condition will be executed. Even if multiple conditions are true, Python stops after executing the first matching block.

In [149]:
age = 25

if age < 13:
  print("You are a child")
elif age < 20:
  print("You are a teenager")
elif age < 65:
  print("You are an adult")
elif age >= 65:
  print("You are a senior")

You are an adult


### Complete If-Elif-Else Chain

temperature = 22

if temperature > 30:
  print("It's hot outside!")
elif temperature > 20:
  print("It's warm outside")
elif temperature > 10:
  print("It's cool outside")
else:
  print("It's cold outside!")

### Python Shorthand If

#### Short Hand If
If you have only one statement to execute, you can put it on the same line as the if statement.

In [154]:
a = 5
b = 2
if a > b: print("a is greater than b")

a is greater than b


#### Short Hand If ... Else
If you have one statement for if and one for else, you can put them on the same line using a conditional expression:

In [157]:
a = 2
b = 330
print("A") if a > b else print("B")

B


This is called a conditional expression (replacement of the "ternary operator").

#### Assign a Value With If ... Else
You can also use a one-line if/else to choose a value and assign it to a variable:

In [158]:
a = 10
b = 20
bigger = a if a > b else b
print("Bigger is", bigger)

Bigger is 20


#### Multiple Conditions on One Line
You can chain conditional expressions, but keep it short so it stays readable:

In [161]:
a = 330
b = 330
print("A") if a > b else print("C") if a == b else print("B")

C


#### Python Logical Operators
Logical operators are used to combine conditional statements. Python has three logical operators:

- and - Returns True if both statements are true
- or - Returns True if one of the statements is true
- not - Reverses the result, returns False if the result is true

In [162]:
a = 200
b = 33
c = 500
if a > b and c > a:
  print("Both conditions are True")

Both conditions are True


## Python Pass Statement

if statements cannot be empty, but if you for some reason have an if statement with no content, put in the pass statement to avoid getting an error.

The pass statement is a null operation - nothing happens when it executes. It serves as a placeholder.

In [164]:
a = 33
b = 200

if b > a:
  pass

#### Why Use pass?

The pass statement is useful in several situations:
- When you're creating code structure but haven't implemented the logic yet
- When a statement is required syntactically but no action is needed
- As a placeholder for future code during development
- In empty functions or classes that you plan to implement later

#### pass in Development

During development, you might want to sketch out your program structure before implementing the details. The pass statement allows you to do this without syntax errors.

Placeholder for future implementation:

In [165]:
age = 16

if age < 18:
  pass # TODO: Add underage logic later
else:
  print("Access granted")

#### pass vs Comments

A comment is ignored by Python, but pass is an actual statement that gets executed (though it does nothing). You need pass where Python expects a statement, not just a comment.

This will cause an error (empty code block):

In [166]:
score = 85

if score > 90:
  # This is excellent
# This will raise an IndentationError

_IncompleteInputError: incomplete input (2330558437.py, line 5)

In [167]:
# This works correctly with pass:

score = 85

if score > 90:
  pass # This is excellent
print("Score processed")

Score processed


#### pass with Multiple Conditions

You can use pass in any branch of an if-elif-else statement.

In [ ]:
value = 50

if value < 0:
  print("Negative value")
elif value == 0:
  pass # Zero case - no action needed
else:
  print("Positive value")

#### pass in Other Contexts

While we focus on pass with if statements here, it's also commonly used with loops, functions, and classes.

In [168]:
def calculate_discount(price):
  pass # TODO: Implement discount logic

# Function exists but doesn't do anything yet

## Python Match

#### The Python Match Statement

Instead of writing many if..else statements, you can use the match statement.

The match statement selects one of many code blocks to be executed.
```py
match expression:
  case x:
    code block
  case y:
    code block
  case z:
    code block
```

This is how it works:

- The match expression is evaluated once.
- The value of the expression is compared with the values of each case.
- If there is a match, the associated block of code is executed.

The example below uses the weekday number to print the weekday name:

In [169]:
day = 4
match day:
  case 1:
    print("Monday")
  case 2:
    print("Tuesday")
  case 3:
    print("Wednesday")
  case 4:
    print("Thursday")
  case 5:
    print("Friday")
  case 6:
    print("Saturday")
  case 7:
    print("Sunday")

Thursday


Use the underscore character _ as the last case value if you want a code block to execute when there are not other matches: </br>
The value _ will always match, so it is important to place it as the last case to make it behave as a default case.

In [171]:
day = 4
match day:
  case 6:
    print("Today is Saturday")
  case 7:
    print("Today is Sunday")
  case _:
    print("Looking forward to the Weekend")

Looking forward to the Weekend


#### Combine Values

Use the pipe character | as an or operator in the case evaluation to check for more than one value match in one case:

In [172]:
day = 4
match day:
  case 1 | 2 | 3 | 4 | 5:
    print("Today is a weekday")
  case 6 | 7:
    print("I love weekends!")

Today is a weekday


#### If Statements as Guards

You can add `if` statements in the case evaluation as an extra condition-check:

In [173]:
month = 5
day = 4
match day:
  case 1 | 2 | 3 | 4 | 5 if month == 4:
    print("A weekday in April")
  case 1 | 2 | 3 | 4 | 5 if month == 5:
    print("A weekday in May")
  case _:
    print("No match")

A weekday in May


## Loops

### Python Loops

Python has two primitive loop commands:
- while loops
- for loops

With the while loop we can execute a set of statements as long as a condition is true.

In [188]:
i = 1
while i < 6:
  print(i, end= " ")
  i += 1

1 2 3 4 5 

The while loop requires relevant variables to be ready, in this example we need to define an indexing variable, i, which we set to 1.

i = 1
while i < 6:
  print(i, end=" ")
  if i == 3:
    break
  i += 1

With the `continue` statement we can stop the current iteration, and continue with the next:

In [194]:
i = 1
while i < 6:
  print(i, end = " ")
  i += 1
else:
  print("\ni is no longer less than 6")

1 2 3 4 5 
i is no longer less than 6


### Python For Loops

A for loop is used for iterating over a sequence (that is either a list, a tuple, a dictionary, a set, or a string).

This is less like the for keyword in other programming languages, and works more like an iterator method as found in other object-orientated programming languages.

With the for loop we can execute a set of statements, once for each item in a list, tuple, set etc.

In [195]:
fruits = ["apple", "banana", "cherry"]
for x in fruits:
  print(x, end = " ")

apple banana cherry 

Even strings are iterable objects, they contain a sequence of characters:

In [196]:
for x in "banana":
  print(x, end = " ")

b a n a n a 

With the break statement we can stop the loop before it has looped through all the items:

In [199]:
fruits = ["apple", "banana", "cherry"]
for x in fruits:
  print(x, end =" ")
  if x == "banana":
    break

apple banana 

With the continue statement we can stop the current iteration of the loop, and continue with the next:

In [201]:
fruits = ["apple", "banana", "cherry"]
for x in fruits:
  if x == "banana":
    continue
  print(x, end = " ")

apple cherry 

### The range() Function
To loop through a set of code a specified number of times, we can use the range() function,

The `range()` function returns a sequence of numbers, starting from 0 by default, and increments by 1 (by default), and ends at a specified number.

In [202]:
for x in range(6):
  print(x, end =" ")

0 1 2 3 4 5 

Note that range(6) is not the values of 0 to 6, but the values 0 to 5.

The `range()` function defaults to `0` as a starting value, however it is possible to specify the starting value by adding a parameter: `range(2, 6)`, which means values from `2` to `6` (but not including 6):

In [203]:
for x in range(2, 6):
  print(x, end = " ")

2 3 4 5 

The `range()` function defaults to increment the sequence by `1`, </br>
however it is possible to specify the increment value by adding a third parameter: `range(2, 30, 3)`:

Increment the sequence with 3 (default is 1):

In [204]:
for x in range(2, 30, 3):
  print(x, end =" ")

2 5 8 11 14 17 20 23 26 29 

#### Else in For Loop

The else keyword in a for loop specifies a block of code to be executed when the loop is finished:

In [209]:
for x in range(6):
  print(x, end =" ")
else:
  print("Finally finished!")

0 1 2 3 4 5 Finally finished!


The else block will NOT be executed if the loop is stopped by a break statement.

Break the loop when `x` is `3`, and see what happens with the else block:

In [211]:
for x in range(6):
  if x == 3: break
  print(x, end =" ")
else:
  print("Finally finished!")

0 1 2 

#### Nested Loops
A nested loop is a loop inside a loop.

The "inner loop" will be executed one time for each iteration of the "outer loop":

In [213]:
adj = ["red", "big", "tasty"]
fruits = ["apple", "banana", "cherry"]

for x in adj:
  for y in fruits:
    print("[{}, {}]". format(x, y), end = " ")

[red, apple] [red, banana] [red, cherry] [big, apple] [big, banana] [big, cherry] [tasty, apple] [tasty, banana] [tasty, cherry] 

### The pass Statement

for loops cannot be empty, but if you for some reason have a for loop with no content, put in the pass statement to avoid getting an error.

In [214]:
for x in [0, 1, 2]:
  pass

## Python Functions

- A function is a block of code which only runs when it is called.
- A function can return data as a result.
- A function helps avoiding code repetition.
- A function is defined using the def keyword, followed by a function name and parentheses:
- The code inside the function must be indented. Python uses indentation to define code blocks.

In [216]:
def my_function():
  print("Hello from a function")

my_function()

Hello from a function


### Function Names

Function names follow the same rules as variable names in Python:
- A function name must start with a letter or underscore
- A function name can only contain letters, numbers, and underscores
- Function names are case-sensitive (myFunction and myfunction are different)

#### Why Use Functions?
Imagine you need to convert temperatures from Fahrenheit to Celsius several times in your program. Without functions, you would have to write the same calculation code repeatedly:

Functions can send data back to the code that called them using the return statement.

When a function reaches a return statement, it stops executing and sends the result back: </br>
If a function doesn't have a return statement, it returns None by default.

In [218]:
def get_greeting():
  return "Hello from a function"

message = get_greeting()
print(message)

Hello from a function


#### The pass Statement

Function definitions cannot be empty. If you need to create a function placeholder without any code, use the pass statement: </br>
The pass statement is often used when developing, allowing you to define the structure first and implement details later.

In [219]:
def my_function():
  pass

### Arguments
Information can be passed into functions as arguments.

Arguments are specified after the function name, inside the parentheses. You can add as many arguments as you want, just separate them with a comma.

The following example has a function with one argument (fname). When the function is called, we pass along a first name, which is used inside the function to print the full name:

In [220]:
def my_function(fname):
  print(fname + " Refsnes")

my_function("Emil")
my_function("Tobias")
my_function("Linus")

Emil Refsnes
Tobias Refsnes
Linus Refsnes


#### Parameters vs Arguments

The terms parameter and argument can be used for the same thing: information that are passed into a function.
From a function's perspective:
- A parameter is the variable listed inside the parentheses in the function definition.
- An argument is the actual value that is sent to the function when it is called.

In [221]:
def my_function(name): # name is a parameter
  print("Hello", name)

my_function("Emil") # "Emil" is an argument

Hello Emil


#### Number of Arguments

By default, a function must be called with the correct number of arguments.</br>
If your function expects 2 arguments, you must call it with exactly 2 arguments.</br>
If you try to call the function with the wrong number of arguments, you will get an error: </br>

In [223]:
def my_function(fname, lname):
  print(fname + " " + lname)

my_function("Emil")

TypeError: my_function() missing 1 required positional argument: 'lname'

#### Default Parameter Values

You can assign default values to parameters. If the function is called without an argument, it uses the default value:

In [224]:
def my_function(name = "friend"):
  print("Hello", name)

my_function("Emil")
my_function("Tobias")
my_function()
my_function("Linus")

Hello Emil
Hello Tobias
Hello friend
Hello Linus


### Keyword Arguments

You can send arguments with the key = value syntax.

In [228]:
def pet(animal, name):
  print("My", animal + "'s name is", name)

pet(animal = "dog", name = "Buddy")

My dog's name is Buddy


This way, with keyword arguments, the order of the arguments does not matter.

In [229]:
pet(name = "Buddy", animal = "dog")

My dog's name is Buddy


### Positional Arguments

When you call a function with arguments without using keywords, they are called positional arguments.

Positional arguments must be in the correct order:

In [232]:
def my_function(animal, name):
  print("My", animal + "'s name is", name)

my_function("dog", "Buddy")
my_function("Buddy", "dog")

My dog's name is Buddy
My Buddy's name is dog


### Mixing Positional and Keyword Arguments

You can mix positional and keyword arguments in a function call.

However, positional arguments must come before keyword arguments:

In [233]:
def my_function(animal, name, age):
  print("I have a", age, "year old", animal, "named", name)

my_function("dog", name = "Buddy", age = 5)

I have a 5 year old dog named Buddy


### Passing Different Data Types
You can send any data type as an argument to a function (string, number, list, dictionary, etc.).

The data type will be preserved inside the function:

Functions can return values using the return statement:

Functions can return any data type, including lists, tuples, dictionaries, and more.

### Positional-Only Arguments

You can specify that a function can have ONLY positional arguments.

To specify positional-only arguments, add , / after the arguments:

In [236]:
def my_function(name, /):
  print("Hello", name)

my_function("Emil")
my_function(name = "Email")

Hello Emil


TypeError: my_function() got some positional-only arguments passed as keyword arguments: 'name'

Without the , / you are actually allowed to use keyword arguments even if the function expects positional arguments:

In [235]:
def my_function(name):
  print("Hello", name)

my_function(name = "Emil")

Hello Emil


### Keyword-Only Arguments

To specify that a function can have only keyword arguments, add *, before the arguments:

In [237]:
def my_function(*, name):
  print("Hello", name)

my_function(name = "Emil")

Hello Emil


Without *,, you are allowed to use positional arguments even if the function expects keyword arguments:

In [238]:
def my_function(name):
  print("Hello", name)

my_function("Emil")

Hello Emil


### Combining Positional-Only and Keyword-Only
You can combine both argument types in the same function.

Arguments before / are positional-only, and arguments after * are keyword-only:

In [239]:
def my_function(a, b, /, *, c, d):
  return a + b + c + d

result = my_function(5, 10, c = 15, d = 20)
print(result)

50


### Python *args and **kwargs

#### *args and **kwargs
By default, a function must be called with the correct number of arguments.

However, sometimes you may not know how many arguments that will be passed into your function.

### *args and **kwargs allow functions to accept a unknown number of arguments.

Arbitrary Arguments - *args
If you do not know how many arguments will be passed into your function, add a * before the parameter name.

This way, the function will receive a tuple of arguments and can access the items accordingly:

Arbitrary Arguments are often shortened to *args in Python documentation.

#### What is *args?
The *args parameter allows a function to accept any number of positional arguments.

Inside the function, args becomes a tuple containing all the passed arguments:

In [240]:
def my_function(*args):
  print("Type:", type(args))
  print("First argument:", args[0])
  print("Second argument:", args[1])
  print("All arguments:", args)

my_function("Emil", "Tobias", "Linus")

Type: <class 'tuple'>
First argument: Emil
Second argument: Tobias
All arguments: ('Emil', 'Tobias', 'Linus')


### Using *args with Regular Arguments
You can combine regular parameters with *args.

Regular parameters must come before *args:

In this example, "Hello" is assigned to greeting, and the rest are collected in names.

In [241]:
def my_function(greeting, *names):
  for name in names:
    print(greeting, name)

my_function("Hello", "Emil", "Tobias", "Linus")

Hello Emil
Hello Tobias
Hello Linus


#### Practical Example with *args
*args is useful when you want to create flexible functions:

In [242]:
def my_function(*numbers):
  total = 0
  for num in numbers:
    total += num
  return total

print(my_function(1, 2, 3))
print(my_function(10, 20, 30, 40))
print(my_function(5))

6
100
5


In [244]:
def max_num(*numbers):
  if len(numbers) == 0:
    return None
  max_num = numbers[0]
  for num in numbers:
    if num > max_num:
      max_num = num
  return max_num

print(max_num(3, 7, 2, 9, 1))

9


### Arbitrary Keyword Arguments - **kwargs

If you do not know how many keyword arguments will be passed into your function, add two asterisks `**` before the parameter name.

This way, the function will receive a dictionary of arguments and can access the items accordingly</br>

Using `**kwargs` to accept any number of keyword arguments </br>
Arbitrary Keyword Arguments are often shortened to `**kwargs` in Python documentation. </br>

In [246]:
def my_function(**kid):
  print("His last name is " + kid["lname"])

my_function(fname = "Tobias", lname = "Refsnes")

His last name is Refsnes


#### What is **kwargs?
The `**kwargs` parameter allows a function to accept any number of keyword arguments.

Inside the function, kwargs becomes a dictionary containing all the keyword arguments:

In [247]:
def my_function(**myvar):
  print("Type:", type(myvar))
  print("Name:", myvar["name"])
  print("Age:", myvar["age"])
  print("All data:", myvar)

my_function(name = "Tobias", age = 30, city = "Bergen")

Type: <class 'dict'>
Name: Tobias
Age: 30
All data: {'name': 'Tobias', 'age': 30, 'city': 'Bergen'}


#### Using **kwargs with Regular Arguments
You can combine regular parameters with **kwargs.

Regular parameters must come before **kwargs:

In [248]:
def my_function(username, **details):
  print("Username:", username)
  print("Additional details:")
  for key, value in details.items():
    print(" ", key + ":", value)

my_function("emil123", age = 25, city = "Oslo", hobby = "coding")

Username: emil123
Additional details:
  age: 25
  city: Oslo
  hobby: coding


#### Combining *args and **kwargs
You can use both *args and **kwargs in the same function.

The order must be:

regular parameters
- *args
- **kwargs

In [249]:
def my_function(title, *args, **kwargs):
  print("Title:", title)
  print("Positional arguments:", args)
  print("Keyword arguments:", kwargs)

my_function("User Info", "Emil", "Tobias", age = 25, city = "Oslo")

Title: User Info
Positional arguments: ('Emil', 'Tobias')
Keyword arguments: {'age': 25, 'city': 'Oslo'}


### Unpacking Arguments

The `*` and `**` operators can also be used when calling functions to unpack (expand) a list or dictionary into separate arguments.

#### Unpacking Lists with *
If you have values stored in a list, you can use `*` to unpack them into individual arguments:

In [250]:
def my_function(a, b, c):
  return a + b + c

numbers = [1, 2, 3]
result = my_function(*numbers) # Same as: my_function(1, 2, 3)
print(result)

6


#### Unpacking Dictionaries with **
If you have keyword arguments stored in a dictionary, you can use `**` to unpack them:

In [251]:
def my_function(fname, lname):
  print("Hello", fname, lname)

person = {"fname": "Emil", "lname": "Refsnes"}
my_function(**person) # Same as: my_function(fname="Emil", lname="Refsnes")

Hello Emil Refsnes


### Python Scope

A variable is only available from inside the region it is created. This is called scope.

#### Local Scope

A variable created inside a function belongs to the local scope of that function, and can only be used inside that function.

A variable created inside a function is available inside that function:

In [1]:
def myfunc():
  x = 300
  print(x)

myfunc()

300


#### Function Inside Function

As explained in the example above, the variable x is not available outside the function, but it is available for any function inside the function:

The local variable can be accessed from a function within the function:

In [2]:
def myfunc():
  x = 300
  def myinnerfunc():
    print(x)
  myinnerfunc()

myfunc()

300


#### Global Scope

A variable created in the main body of the Python code is a global variable and belongs to the global scope.

Global variables are available from within any scope, global and local.

A variable created outside of a function is global and can be used by anyone:

In [3]:
x = 300

def myfunc():
  print(x)

myfunc()

print(x)

300
300


#### Naming Variables

If you operate with the same variable name inside and outside of a function, Python will treat them as two separate variables, one available in the global scope (outside the function) and one available in the local scope (inside the function):

The function will print the local x, and then the code will print the global x:

In [5]:
x = 300

def myfunc():
  x = 200
  print("Inside function", x)

myfunc()

print("In global scope", x)

Inside function 200
In global scope 300


#### Global Keyword

If you need to create a global variable, but are stuck in the local scope, you can use the global keyword.

The global keyword makes the variable global.

If you use the global keyword, the variable belongs to the global scope:

In [6]:
def myfunc():
  global x
  x = 300

myfunc()

print(x)

300


Also, use the global keyword if you want to make a change to a global variable inside a function.

To change the value of a global variable inside a function, refer to the variable by using the global keyword:

In [7]:
x = 300

def myfunc():
  global x
  x = 200

myfunc()

print(x)

200


#### Nonlocal Keyword

The `nonlocal` keyword is used to work with variables inside nested functions.

The `nonlocal` keyword makes the variable belong to the outer function.

If you use the nonlocal keyword, the variable will belong to the outer function:

In [8]:
def myfunc1():
  x = "Jane"
  def myfunc2():
    nonlocal x
    x = "hello"
  myfunc2()
  return x

print(myfunc1())

hello


#### The LEGB Rule

Python follows the LEGB rule when looking up variable names, and searches for them in this order:

- Local - Inside the current function
- Enclosing - Inside enclosing functions (from inner to outer)
- Global - At the top level of the module
- Built-in - In Python's built-in namespace

In [10]:
x = "global"

def outer():
  x = "enclosing"
  def inner():
    x = "local"
    print("Inner:", x)
  inner()
  print("Outer:", x)

outer()
print("Global:", x)

Inner: local
Outer: enclosing
Global: global


### Python Decorators

Decorators let you add extra behavior to a function, without changing the function's code.

A decorator is a function that takes another function as input and returns a new function.

Define the decorator first, then apply it with `@decorator_name` above the function.

By placing `@changecase` directly above the function definition, the function myfunction is being "decorated" with the changecase function.</br>
The function changecase is the decorator.</br>
The function myfunction is the function that gets decorated.

In [13]:
def changecase(func):
  def myinner():
    return func().upper()
  return myinner

@changecase
def myfunction():
  return "Hello Sally"

print(myfunction())

HELLO SALLY


#### Multiple Decorator Calls
A decorator can be called multiple times. Just place the decorator above the function you want to decorate.

Using the `@changecase` decorator on two functions:

In [14]:
def changecase(func):
  def myinner():
    return func().upper()
  return myinner

@changecase
def myfunction():
  return "Hello Sally"

@changecase
def otherfunction():
  return "I am speed!"

print(myfunction())
print(otherfunction())

HELLO SALLY
I AM SPEED!


#### Arguments in the Decorated Function

Functions that requires arguments can also be decorated, just make sure you pass the arguments to the wrapper function </br>
Functions with arguments can also be decorated:

In [16]:
def changecase(func):
  def myinner(x):
    return func(x).upper()
  return myinner

@changecase
def myfunction(nam):
  return "Hello " + nam

print(myfunction("John"))

HELLO JOHN


#### *args and **kwargs

Sometimes the decorator function has no control over the arguments passed from decorated function, to solve this problem, add (`*args`, `**kwargs`) to the wrapper function, this way the wrapper function can accept any number, and any type of arguments, and pass them to the decorated function.

Secure the function with `*args` and `**kwargs` arguments:

In [17]:
def changecase(func):
  def myinner(*args, **kwargs):
    return func(*args, **kwargs).upper()
  return myinner

@changecase
def myfunction(nam):
  return "Hello " + nam

print(myfunction("John"))

HELLO JOHN


#### Decorator With Arguments

Decorators can accept their own arguments by adding another wrapper level. </br>
A decorator factory that takes an argument and transforms the casing based on the argument value.

In [20]:
def changecase(n):
  def changecase(func):
    def myinner():
      if n == 1:
        a = func().lower()
      else:
        a = func().upper()
      return a
    return myinner
  return changecase

@changecase(1)
def lower():
  return "Hello Linus"

@changecase(0)
def higher():
  return "Hello Linus"

print(lower())
print(higher())

print(lower)
print(higher)

hello linus
HELLO LINUS
<function changecase.<locals>.changecase.<locals>.myinner at 0xe29750649850>
<function changecase.<locals>.changecase.<locals>.myinner at 0xe29750649430>


#### Multiple Decorators
You can use multiple decorators on one function.

This is done by placing the decorator calls on top of each other. </br>
Decorators are called in the reverse order, starting with the one closest to the function. </br>
One decorator for upper case, and one for adding a greeting: </br>

In [24]:
def changecase(func):
  def myinner():
    return func().upper()
  return myinner

def addgreeting(func):
  def myinner():
    return "Hello " + func() + " Have a good day!"
  return myinner

@changecase
@addgreeting
def myfunction1(): # first add greeting then change case
  return "Tobias"

print(myfunction1())

@addgreeting
@changecase
def myfunction2(): # first change case then add greeting
  return "Tobias"

print(myfunction2())

HELLO TOBIAS HAVE A GOOD DAY!
Hello TOBIAS Have a good day!


### Preserving Function Metadata

Functions in Python has metadata that can be accessed using the __name__ and __doc__ attributes. </br>
Normally, a function's name can be returned with the __name__ attribute: </br>

In [26]:
def myfunction():
  return "Have a great day!"

print(myfunction.__name__)

myfunction


But, when a function is decorated, the metadata of the original function is lost.

Try returning the name from a decorated function and you will not get the same result:

In [27]:
def changecase(func):
  def myinner():
    return func().upper()
  return myinner

@changecase
def myfunction():
  return "Have a great day!"

print(myfunction.__name__)

myinner


To fix this, Python has a built-in function called functools.wraps that can be used to preserve the original function's name and docstring. </br>
Import functools.wraps to preserve the original function name and docstring.

In [28]:
import functools

def changecase(func):
  @functools.wraps(func)
  def myinner():
    return func().upper()
  return myinner

@changecase
def myfunction():
  return "Have a great day!"

print(myfunction.__name__)

myfunction


## Python Lambda

#### Lambda Functions
A lambda function is a small anonymous function.

A lambda function can take any number of arguments, but can only have one expression. </br>
Syntax `lambda arguments : expression` </br>
The expression is executed and the result is returned

Lambda functions can take any number of arguments:

In [32]:
f = lambda a : a + 10
print(f(5))

f = lambda a, b : a * b
print(f(5, 6))

f = lambda a, b, c : a + b + c
print(f(5, 6, 2))

15
30
13


### Why Use Lambda Functions?
The power of `lambda` is better shown when you use them as an anonymous function inside another function.

Say you have a function definition that takes one argument, and that argument will be multiplied with an unknown number:

In [33]:
def myfunc(n):
  return lambda a : a * n

In [36]:
def myfunc(n):
  return lambda a : a * n

double = myfunc(2)
tripple = myfunc(3)

print(double(11))
print(tripple(11))

22
33


Use lambda functions when an anonymous function is required for a short period of time.

#### Lambda with Built-in Functions
Lambda functions are commonly used with built-in functions like `map()`, `filter()`, and `sorted()`.

Using Lambda with `map()` </br>
The map() function applies a function to every item in an iterable: </br>
Double all numbers in a list:

numbers = [1, 2, 3, 4, 5]
doubled = list(map(lambda x: x * 2, numbers))
print(doubled)

Using Lambda with `filter()` </br>
The filter() function creates a list of items for which a function returns True: </br>
Filter out even numbers from a list:

In [39]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8]
odd_numbers = list(filter(lambda x: x % 2 != 0, numbers))
print(odd_numbers)

[1, 3, 5, 7]


Using Lambda with `sorted()` </br>
The `sorted()` function can use a `lambda` as a `key` for custom sorting:

In [40]:
# Sort a list of tuples by the second element:

students = [("Emil", 25), ("Tobias", 22), ("Linus", 28)]
sorted_students = sorted(students, key=lambda x: x[1])
print(sorted_students)

[('Tobias', 22), ('Emil', 25), ('Linus', 28)]


In [42]:
# Sort strings by length:

words = ["apple", "pie", "banana", "cherry"]
sorted_words = sorted(words, key=lambda x: len(x))
print(sorted_words)

['pie', 'apple', 'banana', 'cherry']


### Python Recursion

#### Recursion
Recursion is when a function calls itself.

Recursion is a common mathematical and programming concept. It means that a function calls itself. This has the benefit of meaning that you can loop through data to reach a result.

The developer should be very careful with recursion as it can be quite easy to slip into writing a function which never terminates, or one that uses excess amounts of memory or processor power. However, when written correctly recursion can be a very efficient and mathematically-elegant approach to programming.

In [48]:
# A simple recursive function that counts down from 5:

def countdown(n):
  if n <= 0:
    print("Done!")
  else:
    print(n, end = " ")
    countdown(n - 1)

countdown(5)

5 4 3 2 1 Done!


#### Base Case and Recursive Case

Every recursive function must have two parts:
- A base case - A condition that stops the recursion
- A recursive case - The function calling itself with a modified argument
Without a base case, the function would call itself forever, causing a stack overflow error.

In [49]:
# Identifying base case and recursive case:

def factorial(n):
  # Base case
  if n == 0 or n == 1:
    return 1
  # Recursive case
  else:
    return n * factorial(n - 1)

print(factorial(5))

120


#### Fibonacci Sequence

The Fibonacci sequence is a classic example where each number is the sum of the two preceding ones. The sequence starts with 0 and 1:

0, 1, 1, 2, 3, 5, 8, 13, ...

The sequence continues indefinitely, with each number being the sum of the two preceding ones.

We can use recursion to find a specific number in the sequence:

In [50]:
# Find the 7th number in the Fibonacci sequence:

def fibonacci(n):
  if n <= 1:
    return n
  else:
    return fibonacci(n - 1) + fibonacci(n - 2)

print(fibonacci(7))

13


#### Recursion with Lists

Recursion can be used to process lists by handling one element at a time:

In [51]:
# Calculate the sum of all elements in a list:

def sum_list(numbers):
  if len(numbers) == 0:
    return 0
  else:
    return numbers[0] + sum_list(numbers[1:])

my_list = [1, 2, 3, 4, 5]
print(sum_list(my_list))

15


In [52]:
# Find the maximum value in a list:

def find_max(numbers):
  if len(numbers) == 1:
    return numbers[0]
  else:
    max_of_rest = find_max(numbers[1:])
    return numbers[0] if numbers[0] > max_of_rest else max_of_rest

my_list = [3, 7, 2, 9, 1]
print(find_max(my_list))

9


#### Recursion Depth Limit

Python has a limit on how deep recursion can go. The default limit is usually around 1000 recursive calls. </br>
Increasing the recursion limit should be done with caution. For very deep recursion, consider using iteration instead. </br>

In [56]:
# Check the recursion limit:

import sys
print(sys.getrecursionlimit())

2000


In [57]:
# If you need deeper recursion, you can increase the limit, but be careful as this can cause crashes:

import sys
sys.setrecursionlimit(2000)
print(sys.getrecursionlimit())

2000


## Python Generators

#### Generators
Generators are functions that can pause and resume their execution.

When a generator function is called, it returns a generator object, which is an iterator.

The code inside the function is not executed yet, it is only compiled. The function only executes when you iterate over the generator.

In [60]:
# A simple generator function:

def my_generator():
  yield 1
  yield 2
  yield 3

for value in my_generator():
  print(value, end = " ")

1 2 3 

Generators allow you to iterate over data without storing the entire dataset in memory.

Instead of using return, generators use the yield keyword.

### The yield Keyword

The `yield` keyword is what makes a function a generator.

When `yield` is encountered, the function's state is saved, and the value is returned. The next time the generator is called, it continues from where it left off.

Unlike return, which terminates the function, yield pauses it and can be called multiple times.

In [63]:
# Generator that yields numbers:

def count_up_to(n):
  count = 1
  while count <= n:
    yield count
    count += 1

for num in count_up_to(5):
  print(num, end = " ")

1 2 3 4 5 

### Generators Saves Memory
Generators are memory-efficient because they generate values on-the-fly instead of storing everything in memory.

For large datasets, generators save memory:

In [65]:
# Generator for large sequences:

def large_sequence(n):
  for i in range(n):
    yield i # program stops here don't go to the next iteration

# This doesn't create a million numbers in memory
gen = large_sequence(1000000)
print(next(gen)) # print the next
print(next(gen))
print(next(gen))

0
1
2


#### Using next() with Generators
You can manually iterate through a generator using the next() function:

In [66]:
def simple_gen():
  yield "Emil"
  yield "Tobias"
  yield "Linus"

gen = simple_gen()
print(next(gen))
print(next(gen))
print(next(gen))

Emil
Tobias
Linus


#### When there are no more values to yield, the generator raises a StopIteration exception:

In [67]:
def simple_gen():
  yield 1
  yield 2

gen = simple_gen()
print(next(gen))
print(next(gen))
print(next(gen)) # This will raise StopIteration

1
2


StopIteration: 

#### Generator Expressions

Similar to list comprehensions, you can create generators using generator expressions with parentheses instead of square brackets:

List comprehension vs generator expression:

In [69]:
# List comprehension - creates a list
list_comp = [x * x for x in range(5)]
print(list_comp)

# Generator expression - creates a generator
gen_exp = (x * x for x in range(5))
print(gen_exp)
print(list(gen_exp))



[0, 1, 4, 9, 16]
<generator object <genexpr> at 0xe297501d41e0>
[0, 1, 4, 9, 16]


#### Using a generator expression with sum:

In [73]:
# Calculate sum of squares without creating a list
total = sum(x * x for x in range(10))
# 1^2 + 2^2 + 3^2 + 4^2 + 5^2 + 6^2 + 7^2 + 8^2 + 9^2
print(total)

285


### Fibonacci Sequence Generator

Generators can be used to create the Fibonacci sequence.

It can continue generating values indefinitely, without running out of memory:

In [78]:
# Generate 100 Fibonacci numbers:

def fibonacci():
  a, b = 0, 1
  while True:
    yield a
    a, b = b, a + b

# Get first 100 Fibonacci numbers
gen = fibonacci()
for _ in range(10):
  print(next(gen), end = " ")

0 1 1 2 3 5 8 13 21 34 

### Generator Methods

Generators have special methods for advanced control:

**send()** Method </br>
The send() method allows you to send a value to the generator:

In [1]:
def echo_generator():
  while True:
    received = yield
    print("Received:", received)

gen = echo_generator()
next(gen) # Prime the generator
gen.send("Hello")
gen.send("World")

Received: Hello
Received: World


**close()** Method </br>
The close() method stops the generator:

In [2]:
def my_gen():
  try:
    yield 1
    yield 2
    yield 3
  except Exception as e:
    print("Exception is", e)
  finally:
    print("Generator closed")

gen = my_gen()
print(next(gen))
gen.close()

1
Generator closed


# Python Modules

### What is a Module?
Consider a module to be the same as a code library.

A file containing a set of functions you want to include in your application.

### Create a Module
To create a module just save the code you want in a file with the file extension `.py`:

#### Example
Save this code in a file named `mymodule.py`
```py
def greeting(name):
  print("Hello, " + name)
```


### Use a Module
Now we can use the module we just created, by using the `import` statement:

#### Example
Import the module named mymodule, and call the greeting function:
```py
import mymodule

mymodule.greeting("Jonathan")
```

> Note: When using a function from a module, use the syntax: module_name.function_name.

### Variables in Module
The module can contain functions, as already described, but also variables of all types (arrays, dictionaries, objects etc):

#### Example
Save this code in the file `mymodule.py`
```py
person1 = {
  "name": "John",
  "age": 36,
  "country": "Norway"
}
```

### Example
Import the module named mymodule, and access the person1 dictionary:
```py
import mymodule

a = mymodule.person1["age"]
print(a)
```

### Naming a Module
You can name the module file whatever you like, but it must have the file extension .py

### Re-naming a Module
You can create an alias when you import a module, by using the as keyword:

#### Example
Create an alias for mymodule called `mx`:
```py
import mymodule as mx

a = mx.person1["age"]
print(a)
```

### Built-in Modules
There are several built-in modules in Python, which you can import whenever you like.

#### Example
Import and use the platform module:
```py
import platform

x = platform.system()
print(x)
```


### Using the dir() Function
There is a built-in function to list all the function names (or variable names) in a module. The `dir()` function:

#### Example
List all the defined names belonging to the platform module:
```py
import platform

x = dir(platform)
print(x)
```
> Note: The `dir()` function can be used on all modules, also the ones you create yourself.

Import From Module
You can choose to import only parts from a module, by using the from keyword.

#### Example
The module named mymodule has one function and one dictionary:
```py
def greeting(name):
  print("Hello, " + name)

person1 = {
  "name": "John",
  "age": 36,
  "country": "Norway"
}
```

#### Example
Import only the person1 dictionary from the module:
```py
from mymodule import person1

print (person1["age"])
```

## Python Datetime

#### Python Dates
A date in Python is not a data type of its own, but we can import a module named `datetime` to work with dates as date objects.

Import the datetime module and display the current date:

In [3]:
import datetime

x = datetime.datetime.now()
print(x)

2025-12-01 12:38:32.951306


When we execute the code from the example above the result will be:

`2025-12-01 12:38:32.951306` </br>
The date contains year, month, day, hour, minute, second, and microsecond. </br>
The `datetime` module has many methods to return information about the date object. </br>
Here are a few examples, you will learn more about them later in this chapter: </br>

Return the year and name of weekday:

In [4]:
import datetime

x = datetime.datetime.now()

print(x.year)
print(x.strftime("%A"))

2025
Monday


### Creating Date Objects
To create a date, we can use the `datetime()` class (constructor) of the datetime module.

The `datetime()` class requires three parameters to create a date: year, month, day. </br>
The `datetime()` class also takes parameters for time and `timezone` (`hour`, `minute`, `second`, `microsecond`, `tzone`), but they are `optional`, and has a default value of `0`, (`None` for `timezone`). </br>

Create a date object:

In [5]:
import datetime

x = datetime.datetime(2020, 5, 17)

print(x)

2020-05-17 00:00:00


### The strftime() Method

The `datetime` object has a method for formatting date objects into readable strings.

The method is called `strftime()`, and takes one parameter, format, to specify the format of the returned string:

Display the name of the month:

In [7]:
import datetime

x = datetime.datetime(2018, 6, 1)

print(x.strftime("%B"))

June


## Python Math

Python has a set of built-in math functions, including an extensive math module, that allows you to perform mathematical tasks on numbers.

### Built-in Math Functions
The `min()` and `max()` functions can be used to find the lowest or highest value in an iterable:

In [9]:
x = min(5, 10, 25)
y = max(5, 10, 25)

print("value is of x", x)
print("value of y", y)

value is of x 5
value of y 25


The abs() function returns the absolute (positive) value of the specified number:

In [12]:
x = -7.25
abs_x = abs(x)

print("abs value of {} is {}".format(x, abs_x))

abs value of -7.25 is 7.25


The `pow(x, y)` function returns the value of `x` to the power of `y`

Return the value of 4 to the power of 3 (same as 4 * 4 * 4):

In [14]:
x = pow(4, 3)
print(x)

64


### Some other functions

In [16]:
import math

# The math.sqrt() method for example, returns the square root of a number:
x = math.sqrt(64)
print(x)

# The math.ceil() method rounds a number upwards to its nearest integer, 
# and the math.floor() method rounds a number downwards to its nearest integer, and returns the result:

x = math.ceil(1.4)
print(x) # returns 2

y = math.floor(1.4)
print(y) # returns 1


x = math.pi
print(x)

8.0
2
1
3.141592653589793


## Python JSON

JSON is a syntax for storing and exchanging data. </br>
JSON is text, written with JavaScript object notation. </br>

### Parse JSON - Convert from JSON to Python
If you have a JSON string, you can parse it by using the json.loads() method.



In [18]:
import json

# some JSON:
x =  '{ "name":"John", "age":30, "city":"New York"}'

# parse x:
y = json.loads(x)

# the result is a Python dictionary:
print(y["age"])

30


### Convert from Python to JSON
If you have a Python object, you can convert it into a JSON string by using the json.dumps() method.

In [19]:
import json

# a Python object (dict):
x = {
  "name": "John",
  "age": 30,
  "city": "New York"
}

# convert into JSON:
y = json.dumps(x)

# the result is a JSON string:
print(y)

{"name": "John", "age": 30, "city": "New York"}


You can convert Python objects of the following types, into JSON strings:
- dict
- list
- tuple
- string
- int
- float
- True
- False
- None

In [20]:
import json

print(json.dumps({"name": "John", "age": 30}))
print(json.dumps(["apple", "bananas"]))
print(json.dumps(("apple", "bananas")))
print(json.dumps("hello"))
print(json.dumps(42))
print(json.dumps(31.76))
print(json.dumps(True))
print(json.dumps(False))
print(json.dumps(None))

{"name": "John", "age": 30}
["apple", "bananas"]
["apple", "bananas"]
"hello"
42
31.76
true
false
null


Convert a Python object containing all the legal data types:

In [21]:
import json

x = {
  "name": "John",
  "age": 30,
  "married": True,
  "divorced": False,
  "children": ("Ann","Billy"),
  "pets": None,
  "cars": [
    {"model": "BMW 230", "mpg": 27.5},
    {"model": "Ford Edge", "mpg": 24.1}
  ]
}

print(json.dumps(x))

{"name": "John", "age": 30, "married": true, "divorced": false, "children": ["Ann", "Billy"], "pets": null, "cars": [{"model": "BMW 230", "mpg": 27.5}, {"model": "Ford Edge", "mpg": 24.1}]}


Pretty-printing with indentation

In [25]:
import json

x = {
  "name": "John",
  "age": 30,
  "married": True,
  "divorced": False,
  "children": ("Ann","Billy"),
  "pets": None,
  "cars": [
    {"model": "BMW 230", "mpg": 27.5},
    {"model": "Ford Edge", "mpg": 24.1}
  ]
}

json_string_pretty = json.dumps(x, indent=4)
print("\nPretty JSON (indent=4):\n")
print(json_string_pretty)


Pretty JSON (indent=4):

{
    "name": "John",
    "age": 30,
    "married": true,
    "divorced": false,
    "children": [
        "Ann",
        "Billy"
    ],
    "pets": null,
    "cars": [
        {
            "model": "BMW 230",
            "mpg": 27.5
        },
        {
            "model": "Ford Edge",
            "mpg": 24.1
        }
    ]
}


Sorting keys for consistent output

In [27]:
import json

x = {
  "name": "John",
  "age": 30,
  "married": True,
  "divorced": False,
  "children": ("Ann","Billy"),
  "pets": None,
  "cars": [
    {"model": "BMW 230", "mpg": 27.5},
    {"model": "Ford Edge", "mpg": 24.1}
  ]
}

json_string_sorted = json.dumps(x, indent=4, sort_keys=True)
print("\nPretty JSON (indent=4, sort_keys=True):")
print(json_string_sorted)


Pretty JSON (indent=4, sort_keys=True):
{
    "age": 30,
    "cars": [
        {
            "model": "BMW 230",
            "mpg": 27.5
        },
        {
            "model": "Ford Edge",
            "mpg": 24.1
        }
    ],
    "children": [
        "Ann",
        "Billy"
    ],
    "divorced": false,
    "married": true,
    "name": "John",
    "pets": null
}


Handling non-ASCII characters

Handling non-ASCII characters

In [28]:
import json

data_non_ascii = {"greeting": "Bonjour à tous!"}
json_string_non_ascii_escaped = json.dumps(data_non_ascii)
print("\nNon-ASCII (escaped by default):")
print(json_string_non_ascii_escaped)

json_string_non_ascii_raw = json.dumps(data_non_ascii, ensure_ascii=False)
print("Non-ASCII (ensure_ascii=False):")
print(json_string_non_ascii_raw)


Non-ASCII (escaped by default):
{"greeting": "Bonjour \u00e0 tous!"}
Non-ASCII (ensure_ascii=False):
{"greeting": "Bonjour à tous!"}


# Python Try Except

The `try` block lets you test a block of code for errors. </br>
The `except` block lets you handle the error. </br>
The `else` block lets you execute code when there is no error. </br>
The `finally` block lets you execute code, regardless of the result of the try- and except blocks. </br>

### Exception Handling
When an error occurs, or exception as we call it, Python will normally stop and generate an error message. </br>
These exceptions can be handled using the try statement: </br>

The `try` block will generate an exception, because the variable is not defined: </br>

Since the try block raises an error, the except block will be executed. </br>
Without the try block, the program will crash and raise an error: </br>

In [29]:
try:
  print(one_undefined_variable)
except:
  print("An exception occurred")

An exception occurred


### Many Exceptions
You can define as many exception blocks as you want, e.g. if you want to execute a special block of code for a special kind of error:

Print one message if the try block raises a NameError and another for other errors:

In [32]:
try:
  print(one_undefined_variable)
except NameError:
  print("Variable one_undefined_variable is not defined")
except:
  print("Something else went wrong")

Variable one_undefined_variable is not defined


### Else
You can use the else keyword to define a block of code to be executed if no errors were raised:

In this example, the try block does not generate any error:

In [34]:
try:
  print("Hello")
except:
  print("Something went wrong")
else:
  print("Nothing went wrong")

Hello
Nothing went wrong


### Finally

The finally block, if specified, will be executed regardless if the try block raises an error or not.

In [36]:
try:
  print(1/0)
except:
  print("Something went wrong")
finally:
  print("The 'try except' is finished")

Something went wrong
The 'try except' is finished


This can be useful to close objects and clean up resources: </br>
Try to open and write to a file that is not writable: </br>
The program can continue, without leaving the file object open. </br>

```py
try:
  f = open("demofile.txt")
  try:
    f.write("Lorum Ipsum")
  except:
    print("Something went wrong when writing to the file")
  finally:
    f.close()
except:
  print("Something went wrong when opening the file")
```

### Raise an exception

As a Python developer you can choose to throw an exception if a condition occurs. </br>
To throw (or raise) an exception, use the `raise` keyword.</br>

Raise an error and stop the program if x is lower than 0:

In [37]:
x = -1

if x < 0:
  raise Exception("Sorry, no numbers below zero")

Exception: Sorry, no numbers below zero

The `raise` keyword is used to raise an exception. </br>
You can define what kind of error to raise, and the text to print to the user. </br>

Raise a TypeError if x is not an integer: 

In [38]:
x = "hello"

if not type(x) is int:
  raise TypeError("Only integers are allowed")

TypeError: Only integers are allowed

## Python User Input

### User Input

Python allows for user input.

That means we are able to ask the user for input.

The following example asks for your name, and when you enter a name, it gets printed on the screen:

In [136]:
print("Enter your name:")
name = input()
print(f"Hello {name}")

Enter your name:


 Abhishek


Hello Abhishek


In the example above, the user had to input their name on a new line. The Python input() function has a prompt parameter, which acts as a message you can put in front of the user input, on the same line:

In [137]:
name = input("Enter your name:")
print(f"Hello {name}")

Enter your name: Abhishe


Hello Abhishe


The input from the user is treated as a string. Even if, in the example above, you can input a number, the Python interpreter will still treat it as a string.

You can convert the input into a number with the float() function:

In [140]:
import math

x = input("Enter a number:")

#find the square root of the number:
y = math.sqrt(float(x))

print(f"The square root of {x} is {y}")
res = math.sqrt(float(x)) * math.sqrt(float(x))
print("Result of sqrt({}) * sqrt({}) is {}".format(x, x, res))

Enter a number: 3


The square root of 3 is 1.7320508075688772
Result of sqrt(3) * sqrt(3) is 2.9999999999999996


### Validate Input

It is a good practice to validate any input from the user. In the example above, an error will occur if the user inputs something other than a number.

To avoid getting an error, we can test the input, and if it is not a number, the user could get a message like "Wrong input, please try again", and allowed to make a new input:

In [142]:
num = None
while not num:
  x = input("Enter a number:")
  try:
    num = float(x);
  except:
    print("Wrong input, please try again.")


Enter a number: as


Wrong input, please try again.


Enter a number: 12
